### PR TITLE
search: add heuristic pattern scanner

### PR DESCRIPTION
**Pull request by rvantonder**
We need a custom scanner that tolerates escape sequences for parentheses and whitespace, which are syntax in our language. This PR is a follow up to the limitation mentioned in https://github.com/sourcegraph/sourcegraph/pull/9620#discussion_r404536378. 

The scanner code is similar to #9727, please see description/rationale for it there.

### CODE REVIEW

File: internal/search/query/parser.go [hunk 1]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
 	return Parameter{Field: "", Value: string(parameter)}
 }
 
+// ScanSearchPattern scans for a pattern using a heuristic that allows it to
```
Comment (tsenart) [inline internal/search/query/parser.go:219]:
```suggestion
// ScanSearchPatternHeuristic scans for a pattern using a heuristic that allows it to
```

File: internal/search/query/parser.go [hunk 2]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+// contain parentheses, if balanced, with appropriate lexical handling for
+// traditional escape sequences, escaped parentheses, and escaped whitespace.
+func ScanSearchPatternHeuristic(buf []byte) ([]string, int, bool) {
+	var count, advance, balanced int
+	var r rune
+	piece := []rune{}
```
Comment (tsenart) [inline internal/search/query/parser.go:225]: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

```suggestion
	var piece []rune
```

File: internal/search/query/parser.go [hunk 3]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+	pieces := []string{}
```
Comment (tsenart) [inline internal/search/query/parser.go:226]:
```suggestion
	var pieces []string
```

File: internal/search/query/parser.go [hunk 4]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+
+	next := func() {
```
Comment (tsenart) [inline internal/search/query/parser.go:228]: Would read a little bit more idiomatic if this closure would return stuff rather mutating variables in its scope.

File: internal/search/query/parser.go [hunk 5]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+loop:
+	for len(buf) > 0 {
+		next()
+		switch {
+		case unicode.IsSpace(r) && balanced == 0:
+			// Stop scanning a potential pattern when we see
+			// whitespace in a balanced state.
+			break loop
+		case r == '(':
+			balanced += 1
```
Comment (tsenart) [inline internal/search/query/parser.go:243]:
```suggestion
			balanced++
```

File: internal/search/query/parser.go [hunk 6]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+			piece = append(piece, r)
+		case r == ')':
+			balanced -= 1
```
Comment (tsenart) [inline internal/search/query/parser.go:246]:
```suggestion
			balanced--
```

File: internal/search/query/parser.go [hunk 7]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+			piece = append(piece, r)
+		case unicode.IsSpace(r):
+			// We see a space and the pattern is unbalanced, so assume this
+			// terminates a piece of an incomplete search pattern.
+			if len(piece) > 0 {
+				pieces = append(pieces, string(piece))
+			}
+			piece = []rune{}
```
Comment (tsenart) [inline internal/search/query/parser.go:254]: Avoid allocation.
```suggestion
			piece = piece[:0]
```

File: internal/search/query/parser.go [hunk 8]
```diff
@@ -216,6 +216,68 @@ func ScanParameter(parameter []byte) Parameter {
...
+		case r == '\\':
+			// Handle escape sequence.
+			if len(buf[advance:]) > 0 {
+				next()
+				switch r {
+				case 'a', 'b', 'f', 'n', 'r', 't', 'v', '\\', '"', '\'', '(', ')':
```
Comment (tsenart) [inline internal/search/query/parser.go:260]: Could this help here? https://golang.org/pkg/unicode/#IsControl

File: internal/search/query/parser.go (hunk 5 above)
Comment (unknwon) [reply internal/search/query/parser.go:243]: I guess this is what happens when we disable the lint, LOL.

Comment (rvantonder) [reply internal/search/query/parser.go:243]: and also when you don't assume this just works for arbitrary languages (thanks Python)

File: internal/search/query/parser.go (hunk 8 above)
Comment (rvantonder) [reply internal/search/query/parser.go:260]: I don't think this is what we want--that function tells whether a rune _is_ a control character, what we care about here is an escape sequence (e.g., in a string) that denotes a control character. Control characters are non-graphic, checking whether one is an escape sequence is just a 'nice thing' we're doing to check that it is one of valid/well-known sequences, but we could choose to reject non-graphic characters and it probably won't make a difference, given our application. The only really important escape sequences are `\\`, `\"`, `\'`, `\(` and `\)`.

Comment (rvantonder) [reply internal/search/query/parser.go:260]: Note that I chose these based on Go's scanner. https://sourcegraph.com/github.com/golang/go@7b90c1c0c42099248fbb849e8a1c3251984688ac/-/blob/src/go/scanner/scanner.go#L553

**Pull request merged**